### PR TITLE
Maintenance: Improve CONTRIBUTORS and its updates

### PR DESCRIPTION
**Pull request by yadij**
The Squid Project used a script to collect CONTRIBUTORS entries, but it
was not called from source-maintenance.sh, it did not really understand
the structure of those entries, and its results required significant
manual polishing efforts. CONTRIBUTORS file kept deteriorating.

This change consists of three major parts detailed further below:

* a major (semi-manual) CONTRIBUTORS update and cleanup;
* scripts/update-contributors.pl: Merges new entries into CONTRIBUTORS;
* collectAuthors() in source-maintenance.sh: Finds new entries.

Part 1: CONTRIBUTORS update: We collected (and then pruned/polished) all
contributors from the following (master and v3+ branches) sources:
    
* all commit authors;
* all commit co-authors (from Co-authored-by trailer and note entries);
* all CONTRIBUTORS file revisions (the latest one was missing entries).

Part 2: update-contributors.pl understands and enforces a more formal
CONTRIBUTORS structure. After a non-indented preamble text ending with
an empty line, indented CONTRIBUTORS entries now use these formats:
    
    name <email>
    name
    <email>

The entries are case-insensitively sorted by the two fields (name,
email), with several conflict-resolution rules aimed at achieving
natural and stable order. Non-ASCII entries are still banned (for now)
because properly supporting them requires a serious dedicated effort.

The program merges current CONTRIBUTORS and all well-formed contributor
entries (read from the standard input) except these:
    
* entries with an already represented email
* entries with an already represented name
* entries containing "squidadm" in name or email

The matching is case-insensitive. These filtering rules appear to work
better in Squid CONTRIBUTORS context than more accurate/strict rules do.

Part 3: collectAuthors() feeds update-contributors.pl with new
contributors collected from "git log" output. The function only looks at
the current git branch commits made after a "vetted" point. That point
is updated by certain CONTRIBUTORS commits, as detailed in the
collectAuthors() description inside source-maintenance.sh. It can also
be specified manually via the new --update-contributors-since option.

It is not critical (and is probably impossible) for CONTRIBUTORS
automation to properly detect and merge every single new contributor.
When the scripts get it wrong, a human can always update CONTRIBUTORS
manually. Rare mistakes are not a big deal in this context. For example,
if a past contributor now needs to be listed with another email (e.g.
representing a new employer), we manually add a second entry.

This change is a reference point for automated CONTRIBUTORS updates.

### CODE REVIEW

File: scripts/source-maintenance.sh
Comment (yadij) [inline scripts/source-maintenance.sh:505]: At this point in the code we need to somewhat cleanup the AUTHORS.tmp list:
* drop all the Squid Project automation accounts (eg squidadm, root)
* de-duplicate core devs accounts from CVS
* rename authors (who have given permission for their details to be updated since initial commit was logged)

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for adding this script. There is one huge problem that we must address. I also would like to use this file-generating PR to define the `CONTRIBUTORS` entry format, adjusting the script accordingly. Details are in the inlined change requests.

File: scripts/source-maintenance.sh
Comment (rousskov) [reply scripts/source-maintenance.sh:509]: > Alex: This approach precludes us from listing contributors not mentioned in git log

No, I was wrong. The script does populate `AUTHORS.tmp` with entries from the existing `CONTRIBUTORS` file. Sorry.

> Perhaps more importantly, prior applications of ~this or~ similar script have resulted in contributor losses

That part is essentially correct, but we will not have that problem going forward, which is a good thing. We just need to restore the already lost entries. Once that bootstrapping is done, the restored entries will be preserved.

File: CONTRIBUTORS
Comment (yadij) [reply CONTRIBUTORS:264]: I think we should. AFAIK, these are UTF-8 and what the author wants to be called.

File: CONTRIBUTORS
Comment (yadij) [reply CONTRIBUTORS:51]: You mean simply placing '#' comments in front of the entries not to be published and filtering them out in the release/snapshot scripts? If so, seems good to me.

File: CONTRIBUTORS
Comment (yadij) [reply CONTRIBUTORS:2]: I propose that your proposal is already the practice. Deviance from it is script problems we need to sort out.

Comment (rousskov) [reply CONTRIBUTORS:2]: I would not call "practice" something that is not actually happening in reality, but I am glad that we seem to agree in principle: We need to adjust the script to use formal entry formats and entry comparison rules for sorting and elimination of duplicates.

File: CONTRIBUTORS
Comment (yadij) [reply CONTRIBUTORS:51]: Folks wanting to hide their involvement are not relevant. "Anonymous" is already an author to give attribution to if someone sends you a patch and ask not to be mentioned. Others provide us with an alias from the outset and do not even let us know their real identity - which is fine too. The data this list is being generated from is already published and publicly available.

The problem to solve is entries which have outdated details or typos from the bzr days when I had to manually a--author everything accepted. eg the 5 entries for me and the 4 for you. The core dev team having old CVS surname-only labels as well as full name+email entries. Basically the author name+email it is too late to go back in history and fix.

Comment (rousskov) [reply CONTRIBUTORS:51]: > Folks wanting to hide their involvement are not relevant

I disagree that participation in an open source project implies _mandatory permanent_ listing in `CONTRIBUTORS`, but we do not need to agree on this issue to make progress -- if you think we should not care about this hypothetical use case, my "deal with it later if needed" suggestion allows us to ignore it for now without further arguments.

> The problem to solve is entries which have outdated details or typos

I hope that my "give `CONTRIBUTORS` special duplicate resolution power" suggestion addresses all of these problems (that we care about). If (and only if) it does not, we will add entry editing capabilities similar to the entry deleting capabilities we just discussed (i.e. `@Edit` instead of `@Silence` and probably using regexes instead of hash keys).

I will experiment and share the results.

Comment (yadij) [reply CONTRIBUTORS:51]: > > Folks wanting to hide their involvement are not relevant
> 
> I disagree that participation in an open source project implies _mandatory permanent_ listing in `CONTRIBUTORS`, 

You misunderstand me.

CONTRIBUTORS is a convenience copy of the " _mandatory permanent_ " VCS history. AFAIK it has two purposes, a) to highlight the appreciation for those contributors and b) for legal analysis of copyrights.

Folks that want to hide their involvement make arrangements to be formally attributed as "Anonymous". So it is clear to those legal folks that the author is *not* you or me etc. We have "Anonymous" in the CONTRIBUTORS already from past occurances of this case. So IMO these folks situation is not relevant to this PR and discussion.


> > The problem to solve is entries which have outdated details or typos
> 
> I agree that this problem exists and should be solved. I hope that my "give `CONTRIBUTORS` special duplicate resolution power" suggestion addresses all of the corresponding problematic cases (that we care about). If (and only if) it does not, we will add entry editing capabilities similar to the entry deleting capabilities we just discussed (i.e. `@Edit` instead of `@Silence` and probably using regexes instead of hash keys).
> 
> I will experiment and share the results.

I occurs to me we might hard-code a commit ID into the script and only auto-include the authors since that commit. Then we could manually filter out the bad entries from CVS/BZR days.

File: CONTRIBUTORS
Comment (yadij) [reply CONTRIBUTORS:2]: My point was it *IS* something that is being done in reality. At least by me for all the non-me patches I have sponsored PRs for.

File: scripts/source-maintenance.sh
Comment (yadij) [reply scripts/source-maintenance.sh:509]: Good find. IIRC that removal was because their contribution was the script which is no longer published. Makes sense to restore the entry though since we are using the whole commit history for CONTRIBUTORS.

File: CONTRIBUTORS
Comment (yadij) [reply CONTRIBUTORS:46]: I have been trying to do that where possible. These empty email brackets have been causing me some trouble with sed patterns and shell syntax.

I have also been contemplating an awk script that assembles an array or name+email to remove those dupes. If you have a good idea please go ahead.

File: CONTRIBUTORS
Comment (rousskov) [reply CONTRIBUTORS]: Resolving since a better idea mentioned in the change request has been implemented. This change request is not the right place to try to agree regarding CONTRIBUTORS effect on copyright and other out-of-scope issues.

File: CONTRIBUTORS
Comment (rousskov) [reply CONTRIBUTORS]: > Would you be willing to polish the script so that it automatically removes obvious duplicates? If not, I will do that.

Done at 9e59207. Resolving this change request. I will describe recent branch changes separately.

**Review by rousskov — COMMENTED** [review]
Resolving/closing a few old change requests after pushing a major update. I will describe those recent changes separately.

File: scripts/source-maintenance.sh
Comment (yadij) [reply scripts/source-maintenance.sh]: Thank you for pushing this forward. And especially for picking up the ```Co-authored-by``` data source I missed.

File: scripts/source-maintenance.sh
Comment (yadij) [reply scripts/source-maintenance.sh:552]: Ack. Bit complicated, but still the best iteration of this updater so far. I am okay with switching to this.

**Review by yadij — COMMENTED** [review]
Above are my review of the [7c43fc0](https://github.com/squid-cache/squid/commit/7c43fc079c19c33f3db35b6f855873db1b02f8b8) changes that should be attended to during manual vetting. (Sorry for the flood, git did not let me make them a proper review set on the non-HEAD commit).

Can you document in ```scripts/update-contributors.pl``` (for lack of a better long-term place) how the vetting is to be done. Particularly the magic phrase (or is it title) to place in the description of future manual vetting PRs.

PS: @rousskov, I will leave for you to approve and pass to Anubis when that is done. AFAIK there is nothing more for me to do in relation to this, though maybe a future PR with vetting updates.

File: CONTRIBUTORS
Comment (rousskov) [inline CONTRIBUTORS:33]: @yadij (or anybody else reading this), if you want to remove or edit some of your own CONTRIBUTORS entries, this branch is a good opportunity to do so. You can edit the in-branch copy of that file (even using GitHub file editing!) at any time before this PR is merged. Thank you.

P.S. Please do not misinterpret my comment as implying that you must remove or change some of your entries! Leaving all of them "as is" is fine.

**Comment by rousskov** [comment]
> Amos: @rousskov, I will leave for you to approve and pass to Anubis when that is done.

> Alex: OK, I will start working on updating the PR branch to match the current master. I will also update the PR description, including adding the magic phrase.

All done.

> Alex: Will clear for merge after that (unless you stop me).

I plan to do that by Monday in case you want to change the PR title/description or polish CONTRIBUTORS further. Please feel free to clear for merging earlier.

**Pull request closed without merging**
